runtime/runtime/src/actions.rs

```diff
@@ -191,10 +191,21 @@ pub(crate) fn action_function_call(
         false,
     );
     let execution_succeeded = match err {
-        Some(VMError::FunctionCallError(err)) => {
-            result.result = Err(ActionErrorKind::FunctionCallError(err).into());
-            false
-        }
+        Some(VMError::FunctionCallError(err)) => match err {
+            FunctionCallError::Nondeterministic(msg) => {
+                panic!("Contract runner returned non-deterministic error '{}', aborting", msg)
+            }
+            FunctionCallError::CompilationError(_)
+            | FunctionCallError::LinkError { msg: _ }
```

Comment:
Unrelated to the PR, but:

* LinkError should probably be a tuple -- it's the only one with `msg` field, which feels inconsistent
* I am not entirely sure we want to distinguish between CompilationError and LinkError, they feel sufficiently similar to me in the context of wasm execution.

Reply:
Guess so, but this PR already become bigger than I want it to be.

Reply:
The intention here is to have error separated between contract executor and the rest of logic. We can make it a enum, but the only purpose of this error is to show error to humans, that's why I made this change.